python/pyspark/ml/pipeline.py

```diff
@@ -89,9 +87,9 @@ def setParams(self, stages=None):
         setParams(self, stages=None)
         Sets params for Pipeline.
         """
-        if stages is None:
-            stages = []
         kwargs = self.setParams._input_kwargs
+        if stages is None:
+            kwargs['stages'] = []
```

Comment:
The logic behind this code
```
if stages is None:
            stages = []
```
is we can not make mutable values as default arguments (you can refer #8110).
So the cause of this bug is we should pass ```[]``` rather than ```None``` to ```self._paramMap[self.stages]```.
I have run the following code:
```
>>> p = Pipeline()
>>> p.fit(training)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yliang/data/trunk1/spark/python/pyspark/ml/base.py", line 64, in fit
    return self._fit(dataset)
  File "/Users/yliang/data/trunk1/spark/python/pyspark/ml/pipeline.py", line 98, in _fit
    for stage in stages:
TypeError: 'NoneType' object is not iterable
>>> p.getStages()
>>> p.setStages([])
Pipeline_40bc8371599065c2f33d
>>> p.getStages()
[]
>>> p.fit(training)
PipelineModel_48a49d414b2eac1eac36
```